ext/ffi/fast_call.rs

```diff
@@ -0,0 +1,1352 @@
+            dynasm!(self.assembler; .arch aarch64; ldr x7, [sp, self.offset_trampoline])
+          }
+        }
+        // 16 and 8 bit integers are 32 bit integers in v8
+        self.offset_trampoline += max(param.size(), 4);
+      }
+
+      (_, param) => {
+        let size_orig = param.size();
+        // 16 and 8 bit integers are 32 bit integers in v8
+        let size_trampl = max(size_orig, 4);
+        // https://developer.apple.com/documentation/xcode/writing-arm64-code-for-apple-platforms:
+        // > Function arguments may consume slots on the stack that are not multiples of 8 bytes.
+        // > If the total number of bytes for stack-based arguments is not a multiple of 8 bytes,
+        // > insert padding on the stack to maintain the 8-byte alignment requirements.
+        // TODO: V8 does not currently follow this Apple's policy, and instead aligns all arguments to 8 Byte boundaries.
+        // A decision needs to be taken:
+        // 1. leave it broken and wait for v8 to fix the bug
+        // 2. Adapt to v8 bug and follow its ABI instead of Apple's. When V8 fixes the implementation, we'll have to fix it here as well
+        let padding_trampl =
+          (size_trampl - self.offset_trampoline % size_trampl) % size_trampl;
+        let padding_orig =
+          (size_orig - self.offset_original % size_orig) % size_orig;
+        match param {
+          Signed(B) | Unsigned(B) => dynasm!(self.assembler
+            ; .arch aarch64
+            ; ldr w8, [sp, self.offset_trampoline + padding_trampl]
+            ; strb w8, [sp, self.offset_original + padding_orig]
+          ),
+          Signed(W) | Unsigned(W) => dynasm!(self.assembler
+            ; .arch aarch64
+            ; ldr w8, [sp, self.offset_trampoline + padding_trampl]
+            ; strh w8, [sp, self.offset_original + padding_orig]
+          ),
+          Signed(DW) | Unsigned(DW) => dynasm!(self.assembler
+            ; .arch aarch64
+            ;  ldr w8, [sp, self.offset_trampoline + padding_trampl]
+            ; str w8, [sp, self.offset_original + padding_orig]
+          ),
+          Signed(QW) | Unsigned(QW) => dynasm!(self.assembler
+            ; .arch aarch64
+            ; ldr x8, [sp, self.offset_trampoline + padding_trampl]
+            ; str x8, [sp, self.offset_original + padding_orig]
+          ),
+        }
+        self.offset_trampoline += padding_trampl + size_trampl;
+        self.offset_original += padding_orig + size_orig;
+      }
+    };
+  }
+
+  fn zero_first_arg(&mut self) {
+    dynasm!(self.assembler
+      ; .arch aarch64
+      ; mov x0, xzr
+    );
+  }
+
+  fn tailcall(&mut self, ptr: *const c_void) {
+    // stack pointer is never modified and remains aligned
+    // frame pointer remains the one provided by the trampoline's caller (V8)
+
+    let mut address = ptr as u64;
+    let mut imm16 = address & 0xFFFF;
+    dynasm!(self.assembler
+      ; .arch aarch64
+      ; movz x8, imm16 as u32
```

Comment:
https://dinfuehr.github.io/blog/encoding-of-immediate-values-on-aarch64/

Reply:
Actually, I'm curious as to why there's `as u32` as opposed to `as u16`?

Reply:
> Actually, I'm curious as to why there's as u32 as opposed to as u16?

That's a Dynasm implementation quirk. This macro expands to

```rust
self.assembler
        .push_u32(3531603976u32 | ((imm16 as u32 & 65535u32) << 5u8));
```

Being `imm16 as u32` the expression given by the user. Technically, adding an `as u32` to force-cast the expression into a u32 (in [here](https://github.com/CensoredUsername/dynasm-rs/blob/9344c87cdb6ab29b1e3ac36333aa07948bc3945c/plugin/src/arch/aarch64/compiler.rs#L177-L179)) would be enough, but it is a design choice of Dynasm not to perform any implicit coercion

Reply:
> Being `imm16 as u32` the expression given by the user. Technically, adding an `as u32` to force-cast the expression into a u32 (in [here](https://github.com/CensoredUsername/dynasm-rs/blob/9344c87cdb6ab29b1e3ac36333aa07948bc3945c/plugin/src/arch/aarch64/compiler.rs#L177-L179)) would be enough, but it is a design choice of Dynasm not to perform any implicit coercion

It's weird because I would expect the macro to accept a u16, and internally upcast that - but I'll raise that upstream...